// @framer/hyfive.weather-ui/code/FormAddPhoto.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";

const style: React.CSSProperties = {
  height: "100%",
  display: "flex",
  color: "#8855FF",
  background: "white",
  overflow: "hidden"
};

// Define type of property
interface Props {
  text: string;
  sample: string;
  shorter: boolean;
}

export class FormAddPhoto extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    sample: "",
    width: 375,
    height: 76,
    shorter: false,

  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    text: { type: ControlType.String, title: "Text" },
    sample: {type: ControlType.Image, title: "Sample Photo"},
    shorter: {type: ControlType.Boolean, title: "Shorter"},
  };

  render() {

    var samplePhoto, iconAddPhoto;
    if(this.props.sample){
        samplePhoto = styles.samplePhoto;
    } else {
        samplePhoto = styles.noSample;
    }


    if(this.props.shorter == true){
        iconAddPhoto = styles.iconAddPhotoWithIcon;
    } else {
        iconAddPhoto = styles.iconAddPhoto;
    }

    return (
      <div style={style}>
        <div style={iconAddPhoto}>
          <svg xmlns="http://www.w3.org/2000/svg" width="60" height="60">
            <path
              d="M 8 1 C 4.134 1 1 4.134 1 8 L 1 52 C 1 55.866 4.134 59 8 59 L 52 59 C 55.866 59 59 55.866 59 52 L 59 8 C 59 4.134 55.866 1 52 1 Z M 8 0 L 52 0 C 56.418 0 60 3.582 60 8 L 60 52 C 60 56.418 56.418 60 52 60 L 8 60 C 3.582 60 0 56.418 0 52 L 0 8 C 0 3.582 3.582 0 8 0 Z M 29.2 29.2 L 17.2 29.2 L 17.2 30.8 L 29.2 30.8 L 29.2 42.8 L 30.8 42.8 L 30.8 30.8 L 42.8 30.8 L 42.8 29.2 L 30.8 29.2 L 30.8 17.2 L 29.2 17.2 Z"
              fill="rgba(0,0,0,.2)"
            />
          </svg>
        </div>
        <div style={samplePhoto}><img style ={ styles.sample } src={this.props.sample} alt="Sample"/></div>
      </div>
    );
  }
}

const styles = {
    iconAddPhoto:{
        margin: "12px 12px 4px 20px",   
    },

    iconAddPhotoWithIcon: {
        margin: "12px 12px 4px 60px",
    },

    samplePhoto:{
        width: 60, height: 60,
        marginTop: 12, 
        borderRadius: 8,
        overflow: "hidden",
    },

    sample:{
        height: "auto",
        width: "100%",
    },
    noSample:{
        display: "none",
    },
}
