// @framer/hyfive.weather-ui/code/FormTextInput.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";

const style: React.CSSProperties = {
  height: "100%",
  display: "flex",
  alignItems: "center",
  //   justifyContent: "center",
  //   textAlign: "center",
  //   background: "rgba(136, 85, 255, 0.1)",
  overflow: "hidden",
  background: "white",
};

// Define type of property
interface Props {
  text: string;
  shorter: Boolean;
  textColor: string;
}

export class FormTextField extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    text: "Explainer text",
    textColor: "light",
    shorter: false,
    width: 375,
    height: 58,
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    text: { type: ControlType.String, title: "Text" },
    textColor: {
      type: ControlType.Enum,
      options: ["light", "dark"],
      optionTitles: ["Light", "Dark"],
      title: "Text Color"
    },
    shorter: { type: ControlType.Boolean, title: "Shorter" }
  };

  render() {
    var textField, textColor;
    if (this.props.shorter == false) {
      textField = styles.textField;
    } else {
      textField = styles.shorterTextField;
    }

    if (this.props.textColor == "light") {
      textColor = { color: "rgba(0,0,0,.38)" };
    } else {
      textColor = { color: "rgba(0,0,0,.87)" };
    }

    return (
      <div style={style}>
        <div style={textField}>
          <span style={textColor}>{this.props.text}</span>
        </div>
      </div>
    );
  }
}

const styles = {
  textField: {
    fontSize: 16,
    lineHeight: 1.5,
    width: "100%",
    padding: "14px 0px",
    margin: "2px 20px",
    borderBottom: "1px solid rgba(0,0,0,.1)"
  },

  textFieldWithInput: {
    fontSize: 16,
    lineHeight: 1.5,
    width: "100%",
    padding: "14px 0px",
    margin: "2px 20px",
    // color: "rgba(0,0,0,.87)",
    borderBottom: "1px solid rgba(0,0,0,.1)"
  },

  shorterTextField: {
    fontSize: 16,
    lineHeight: 1.5,
    width: "100%",
    padding: "14px 0px",
    margin: "2px 20px 2px 60px",
    borderBottom: "1px solid rgba(0,0,0,.1)"
  }
};
