// @framer/hyfive.weather-ui/code/FormOpenTime.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";

const style: React.CSSProperties = {
  height: "100%",
  display: "flex",
  overflow: "visible",
  fontSize: 16,
  background: "white",
};

// Define type of property
interface Props {
  day: string;
  time: string;
  state: string;
  shorter: boolean,

}

export class FormOpenTime extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    day: "每天",
    time: "08:00 - 20:00",
    state: "add",
    shorter: false,
    width: 375,
    height: 57
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    day: {
      type: ControlType.Enum,
      options: ["每天", "工作日", "节假日"],
      title: "Day"
    },
    time: { type: ControlType.String, title: "Time" },
    state:{ 
        type: ControlType.SegmentedEnum,
        options: ["add","remove"],
        optionTitles:["+","-"],
        title: "State",
    },
    shorter:{
      type: ControlType.Boolean,
      title: "Shorter",
    }

  };

  render() {

    var stateAdd, stateRemove,wrapper;
    if(this.props.state=="add"){
        stateAdd = styles.stateShow;
        stateRemove = styles.stateHide;
    } else{
        stateAdd = styles.stateHide;
        stateRemove = styles.stateShow;
    }

    if(this.props.shorter == true){
      wrapper = styles.wrapperShorter;
    } else {
      wrapper = styles.wrapper;
    }

    return (
      <div style={style}>
        <div style={wrapper}>
          <div style={styles.day}>{this.props.day}</div>
          <div style={styles.time}>{this.props.time}</div>
          <div style={stateAdd}>
            <svg xmlns="http://www.w3.org/2000/svg" width="22" height="22">
              <path
                d="M 10.143 10.143 L 10.143 3.857 C 10.143 3.384 10.527 3 11 3 C 11.473 3 11.857 3.384 11.857 3.857 L 11.857 10.143 L 18.143 10.143 C 18.616 10.143 19 10.527 19 11 C 19 11.473 18.616 11.857 18.143 11.857 L 11.857 11.857 L 11.857 18.143 C 11.857 18.616 11.473 19 11 19 C 10.527 19 10.143 18.616 10.143 18.143 L 10.143 11.857 L 3.857 11.857 C 3.384 11.857 3 11.473 3 11 C 3 10.527 3.384 10.143 3.857 10.143 Z"
                fill="rgba(0,0,0,1)"
              />
            </svg>
          </div>
          <div style = {stateRemove}>
            <svg xmlns="http://www.w3.org/2000/svg" width="22" height="22">
              <path
                d="M 18.143 10.143 C 18.616 10.143 19 10.527 19 11 C 19 11.473 18.616 11.857 18.143 11.857 L 3.857 11.857 C 3.384 11.857 3 11.473 3 11 C 3 10.527 3.384 10.143 3.857 10.143 Z"
                fill="rgba(0,0,0,1)"
              />
            </svg>
          </div>
        </div>
      </div>
    );
  }
}

const styles = {
  day: {
    display: "inline-flex",
    //   justifyContent: "center",
    alignItems: "center",
    height: "100%",
    width: 60,
    borderBottom: "1px solid rgba(0,0,0,.1)"
  },

  time: {
    display: "inline-flex",
    // justifyContent: "center",
    alignItems: "center",
    height: "100%",
    width: 130,
    marginLeft: 20,
    borderBottom: "1px solid rgba(0,0,0,.1)"
  },

  wrapper: {
    display: "inline-flex",
    height: 50,
    alignItems: "center",
    width: "100%",
    margin: "4px 20px"
    //   background: "aqua",
  },

  wrapperShorter:{
    display: "inline-flex",
    height: 50,
    alignItems: "center",
    width: "100%",
    margin: "4px 20px 4px 60px"
  },

  stateShow:{
    position: "absolute",
    right: 20,
  },

  stateHide:{
    display: "none",
  },
};
