// @framer/hyfive.weather-ui/code/HeaderWithSubtitle.tsx
import * as React from "react";
import { PropertyControls, ControlType, Data } from "framer";
import { Header } from "./Header";
import "../styles/style-index.css";



// Define type of property
interface Props {
  headerSize: string;
  header: string;
  subtitle: string;
  transparent: boolean;


}

export class Header_with_Subtitle extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    header: "Feather UI",
    headerSize: "h1",
    subtitle: "Make wings, hence fly",
    width: 375,
    height: 90,
    transparent: false,
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {

    headerSize: {
      type: ControlType.SegmentedEnum,
      title: "Size",
      options: ["h1", "h2", "h3", "h4", "h5"],
      optionTitles: ["H1", "H2", "H3", "H4", "H5"]
    },
    header: { type: ControlType.String, title: "Header" },
    subtitle: { type: ControlType.String, title: "Subtitle" },
    transparent: { type: ControlType.Boolean, title: "Transparent" },

  };

  render() {

    if (this.props.transparent == false) {
      stColor = "var(--color-secondary)";
    } else {
      stColor = "white";
    }



    const styles = {
      style: {
        display: "flex",
        flexDirection: "column",
        color: "#8855FF",
        overflow: "visible",
        background: "white",
    
      },
    
      styleTS: {
        display: "flex",
        flexDirection: "column",
        color: "white",
        overflow: "visible",
    
      },
    
    
        h1: {
        padding: "0px 20px 4px",
        color: stColor,
        lineHeight: 1.5,
        fontWeight: 200,
        fontSize: 16,
      },
    
      h2: {
        padding: "2px 20px 4px",
        color: "#878787",
        lineHeight: 1.5,
        fontWeight: 200,
        fontSize: 14,
      },
    
      h3: {
        padding: "2px 20px 4px",
        color: "#878787",
        lineHeight: 1.5,
        fontWeight: 200,
        fontSize: 14,
      },
    
      h4: {
        padding: "2px 20px 4px",
        color: "#878787",
        lineHeight: 1.5,
        fontWeight: 200,
        fontSize: 14,
      },
    
      h5: {
        padding: "4px 20px 4px",
        color: "#878787",
        lineHeight: 1.4,
        fontWeight: 200,
        fontSize: 12,
      },
    }

    var stylesStyle : React.CSSProperties = {
     
        display: "flex",
        flexDirection: "column",
        color: "#8855FF",
        overflow: "visible",
        background: "green"
    
  
    }
    

    var styleST, style, stColor;

    styleST = styles[this.props.headerSize];

    if (this.props.transparent == false) {
      style = styles.style;
    } else {
      style = styles.styleTS;
    }

    return (

      <div style={style}>
        <Header text={this.props.header} size={this.props.headerSize} />
        <div style={styleST}> {this.props.subtitle}</div>
      </div>
    );
  }
}
