// @framer/hyfive.weather-ui/code/Header.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import "../styles/shared.css";
import "../styles/style-index.css";

// Define type of property
interface Props {
  text: string;
  size: string;
  width: number;
  height: number;
}


export class Header extends React.Component<Partial<Props>> {
  // Set default properties
  static defaultProps = {
    text: "Feather UI",
    size: "h1",
    width: 375,
    height: 62,
  };

  // Items shown in property panel
  static propertyControls = {
    text: { type: ControlType.String, title: "Text" },
    size: {
      type: ControlType.SegmentedEnum,
      title: "Size",
      options: ["h1", "h2", "h3", "h4", "h5"],
      optionTitles: ["H1", "H2", "H3", "H4", "H5"]
    }
  };

  render() {
    var style;

    const width = 375;
    const height = 90;

    style = styles[this.props.size];
    
    return (
      <div className ="headerSharedStyle" style={style} >
        {this.props.text}
      </div>
    );
  }
}



const styles = {
 
  "h1": {
    fontSize: "var(--h1)",
    fontWeight: 800,
  },
  "h2": {
    fontSize: "var(--h2)",
  },
  "h3": {
    fontSize: "var(--h3)",
  },
  "h4": {
    fontSize: "var(--h4)",
  },
  "h5": {
    fontSize: "var(--h5)",
  }
};
