// @framer/hyfive.weather-ui/code_/FormHeaderWithSubtitle.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import { FormHeader } from "./FormHeader";

const style: React.CSSProperties = {
  display: "flex",
  flexDirection:"column",
  overflow: "visible",
  background: "white",
};

// Define type of property
interface Props {
  headerText: string;
  required: boolean;
  withIcon: boolean;
  subtitle: string;
}

export class FormHeaderWithSubtitle extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    headerText: "Header",
    subtitle: "Subtitle goes here :)",
    required: false,
    withIcon: false,
    width: 375,
    height: 88,
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    headerText: { type: ControlType.String, title: "Header Text" },
    subtitle: { type: ControlType.String, title: "Subtitle" },
    required: { type: ControlType.Boolean, title: "Required" },
    withIcon: { type: ControlType.Boolean, title: "Icon" }
  };

  render() {


    var subtitle;

    if(this.props.withIcon == true){
        subtitle = styles.subtitleWithIcon;
    } else {
        subtitle = styles.subtitle;
    }

    return (
      <div style={style}>
        <FormHeader
          text={this.props.headerText}
          required={this.props.required}
          withIcon={this.props.withIcon}
          children = {this.props.children}
        />
        <div style={subtitle}> {this.props.subtitle}</div>
      </div>
    );
  }
}

const styles = {
    subtitle:{
        padding: "4px 20px 4px",
        color:"#878787",
        lineHeight: 1.4,
        fontWeight: 400,
        fontSize: 14,
    },

    subtitleWithIcon:{
        padding: "4px 20px 4px 62px",
        color:"#878787",
        lineHeight: 1.4,
        fontWeight: 400,
        fontSize: 14,
    }
}
