// @framer/rmarinac89.ios-tab-bar/code/HomeIndicator.tsx
import * as React from "react";
import { PropertyControls, ControlType } from "framer";
import styled from "styled-components";

const IndicatorWrapper = styled.div`
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
`;

// Define type of property
interface Props {
  height: number;
  width: number;
  tint: "dark" | "light";
  // device: "iPhone X" | "iPad Pro";
}

export class HomeIndicator extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    height: 34,
    width: 375,
    tint: "dark"
  };

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    tint: {
      type: ControlType.Enum,
      options: ["dark", "light"],
      optionTitles: ["Dark", "Light"],
      title: "Tint"
    }
  };

  render() {
    const { tint } = this.props;
    const isDark = tint === "dark";
    const backgroundColor = isDark ? "#000000" : "#FFFFFF";

    let indicatorWidth = "134px";

    if (this.props.width > 375 && this.props.width <= 812) {
      indicatorWidth = "209px";
    } else if (this.props.width > 834 && this.props.width <= 1024) {
      indicatorWidth = "273px";
    } else if (this.props.width >= 1194 && this.props.width <= 1366) {
      indicatorWidth = "315px";
    } else {
      indicatorWidth = "35.7%";
    }

    const IndicatorBar = styled.div`
      background-color: ${backgroundColor};
      height: 5px;
      border-radius: 50px;
      width: ${indicatorWidth};
      margin: auto;
      margin-top: 21px;
    `;

    return (
      <IndicatorWrapper>
        <IndicatorBar />
      </IndicatorWrapper>
    );
  }
}

export default HomeIndicator;
