// @framer/rmarinac89.ios-tab-bar/code/TabBar.tsx
import * as React from "react"
import { PropertyControls, ControlType } from "framer"
import { Icon } from "./IconGenerator"
import styled from "styled-components"

const TabIcon = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  padding-top: 4px;
  font-size: 10px;
  align-items: center;
  background: transparent;
  height: 48px;
  width: 49px;
`

const TabLabel = styled.p`
  margin: 0;
  flex: 0 100%;
  text-align: center;
  font-weight: 500;
`

// Define type of property
interface Props {
    Tabs: number
    labelOne: string
    labelTwo: string
    labelThree: string
    labelFour: string
    labelFive: string
    IconOne: string
    IconTwo: string
    IconThree: string
    IconFour: string
    IconFive: string
    labels: boolean
    height: number
    width: number
    activeTab: number
    inactiveColor: string
    activeColor: string
    tabBackgroundColor: string
    set: "feather" | "material" | "ionicons" | "fontawesome" | "custom"
}

export class TabBar extends React.Component<Props> {
    // Set default properties
    static defaultProps = {
        Tabs: 5,
        labelOne: "Home",
        labelTwo: "Search",
        labelThree: "Inbox",
        labelFour: "Archives",
        labelFive: "Settings",
        labels: true,
        IconOne: "home",
        customIconOne: " ",
        customIconTwo: " ",
        customIconThree: " ",
        customIconFour: " ",
        customIconFive: " ",
        IconTwo: "search",
        IconThree: "inbox",
        IconFour: "folder",
        IconFive: "settings",
        height: 83,
        width: 375,
        inactiveColor: "#A0A1A0",
        activeColor: "#0055FF",
        activeTab: 1,
        tabBackgroundColor: "rgba(250, 250, 250, 0.9)",
        set: "feather",
    }

    // Items shown in property panel
    static propertyControls: PropertyControls = {
        set: {
            type: ControlType.Enum,
            title: "Icon Set",
            options: ["feather", "material", "ionicons", "fontawesome"],
            optionTitles: ["Feather", "Material", "Ionicons", "Font Awesome"],
        },
        labels: {
            type: ControlType.Boolean,
            title: "Labels",
            disabledTitle: "Hide",
            enabledTitle: "Show",
        },
        Tabs: {
            type: ControlType.Number,
            min: 2,
            max: 5,
            title: "Tabs",
            displayStepper: true,
        },
        activeTab: {
            type: ControlType.Number,
            min: 1,
            max: 5,
            title: "Active Tab",
        },
        activeColor: { type: ControlType.Color, title: "Active Color" },
        inactiveColor: { type: ControlType.Color, title: "Inactive Color" },
        tabBackgroundColor: { type: ControlType.Color, title: "Background" },
        IconOne: {
            type: ControlType.String,
            title: "Icon 1",
            placeholder: "Icon Name",
        },
        labelOne: {
            type: ControlType.String,
            title: "Label 1",
            placeholder: "Label Text",
            hidden: props => props.labels !== true,
        },
        IconTwo: {
            type: ControlType.String,
            title: "Icon 2",
            placeholder: "Icon Name",
            hidden: props => props.Tabs < 2,
        },

        labelTwo: {
            type: ControlType.String,
            title: "Label 2",
            placeholder: "Label Text",
            hidden: props => props.labels !== true || props.Tabs < 2,
        },
        IconThree: {
            type: ControlType.String,
            title: "Icon 3",
            placeholder: "Icon Name",
            hidden: props => props.Tabs < 3,
        },

        labelThree: {
            type: ControlType.String,
            title: "Label 3",
            placeholder: "Label Text",
            hidden: props => props.labels !== true || props.Tabs < 3,
        },
        IconFour: {
            type: ControlType.String,
            title: "Icon 4",
            placeholder: "Icon Name",
            hidden: props => props.Tabs < 4,
        },

        labelFour: {
            type: ControlType.String,
            title: "Label 4",
            placeholder: "Label Text",
            hidden: props => props.labels !== true || props.Tabs < 4,
        },
        IconFive: {
            type: ControlType.String,
            title: "Icon 5",
            placeholder: "Icon Name",
            hidden: props => props.Tabs < 5,
        },

        labelFive: {
            type: ControlType.String,
            title: "Label 5",
            placeholder: "Label Text",
            hidden: props => props.labels !== true || props.Tabs < 5,
        },
    }

    render() {
        const {
            Tabs,
            labels,
            labelOne,
            labelTwo,
            labelThree,
            labelFour,
            labelFive,
            IconOne,
            IconTwo,
            IconThree,
            IconFour,
            IconFive,
            inactiveColor,
            activeColor,
            activeTab,
            tabBackgroundColor,
            set,
            width,
        } = this.props

        const StackGroup = styled.div`
      display: flex;
      justify-content: space-around;
      height: 100%;
      width: 100%;
      background: ${tabBackgroundColor};
      box-shadow: 0 0.5px 0 0 rgba(0, 0, 0, 0.25) inset;
    `

        var items = []
        var tabItems = [
            [labelOne, IconOne],
            [labelTwo, IconTwo],
            [labelThree, IconThree],
            [labelFour, IconFour],
            [labelFive, IconFive],
        ]

        for (var index = 0; index < Tabs; index++) {
            const checkColor =
                activeTab === index + 1 ? activeColor : inactiveColor
            items.push(
                <TabIcon
                    style={width > 767 ? { width: "100%" } : { width: 49 }}
                    key={index}
                >
                    <Icon
                        icon={tabItems[index][1]}
                        width={26}
                        height={26}
                        color={checkColor}
                        set={set}
                    />
                    <TabLabel
                        style={
                            width > 767
                                ? {
                                      textAlign: "left",
                                      flex: 0,
                                      marginLeft: "10px",
                                      color: checkColor,
                                  }
                                : {
                                      textAlign: "center",
                                      flex: "0 100%",
                                      marginLeft: "0",
                                      color: checkColor,
                                  }
                        }
                    >
                        {labels ? tabItems[index][0] : " "}
                    </TabLabel>
                </TabIcon>
            )
        }

        return <StackGroup>{items}</StackGroup>
    }
}
